### DISCUSSION TITLE
ESP32-S3 Intermittent 30ms I2C Delay

### DISCUSSION THREAD

**Discussion by EricNRS** [zephyrproject-rtos/zephyr#63669] [General] [answered]
I am seeing an I2C timeout when making a series of I2C requests to collect the voltage and current from INA237 current sensors.

Whenever the timeout occurs, it stalls the transfer for 29 to 30 milliseconds which is an issue with a 100ms update loop and around 24 I2C transactions during this time.  The SDA line is high and the SCL line is low.  I added a resistor between the ESP32-S3 and the INA237 it proved that the ESP32-S3 is holding the SCL line low.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/127365838/3afe3790-eeaf-4b24-9a43-fc2029dc642b)

I added debug messages to i2c_esp32_transmit, and the delay is when waiting for the ISR to give the `cmd_sem`.
https://github.com/zephyrproject-rtos/zephyr/blob/a6adaedd51ffe1b658304da798892bce0a67382b/drivers/i2c/i2c_esp32.c#L325

I added delay checks to `i2c_esp32_isr` for the calls to `i2c_hal_master_handle_tx_event()` and `i2c_hal_master_handle_rx_event()` and neither were the issue, so the problem seems to be a delay in the ESP32-S3 HAL in processing the event and calling the ISR.
https://github.com/zephyrproject-rtos/zephyr/blob/a6adaedd51ffe1b658304da798892bce0a67382b/drivers/i2c/i2c_esp32.c#L670

Key question is does the 30ms delay match anything in the HAL code?  That might provide a clue as to where the issue is located.  I have been unable to find it after spending several hours looking through the code.

**Comment by EricNRS** [comment]
@sylvioalves Could you please comment on this when you are back in the office on Monday?

Cheers,
Eric

**Reply by sylvioalves** [reply]
@EricNRS, do you have this merged in this test or not? https://github.com/zephyrproject-rtos/zephyr/pull/63149

**Reply by EricNRS** [reply]
Good question.  My branch was about 3 weeks old and was missing the change, but cherry-picking the change to my branch did not make any difference. 

Any ideas why there would be a 30ms delay?  That delay is always 29.9 to 30ms, so seems like a timeout in the I2C state machine somewhere.  I wonder if there is a race condition leading to a missing event.  I tried doing a 50ms delay between each read to see if the issue would go away, but it was still there.  Very strange.

Here is the read operation that is being done:
https://github.com/zephyrproject-rtos/zephyr/blob/a6adaedd51ffe1b658304da798892bce0a67382b/drivers/sensor/ina23x/ina23x_common.c#L26-L39

Right now I do not have a test case for it to allow you to reproduce the issue.  Let me know if you have any ideas.

**Reply by EricNRS** [reply]
I did some more testing here and the problem shows up for single-byte write-read reads as well.

This code reproduces the issue for me for a 6408-compatible IO Expander.  The expander doesn't have timeouts, so the read is just delayed (my previous testing with the INA237 delays shows all 1's after the 30ms delay which may be an internal timeout of the INA237).  I call this function 8 times in a loop every 100ms and the issue gets reproduced within 1 minute.

```c
		struct i2c_dt_spec i2c_device = I2C_DT_SPEC_GET(DT_NODELABEL(status_leds));
		uint8_t io6408_id_register = 0x01;
		uint8_t value;
		uint32_t start_time, end_time;
		int ret;

		start_time = k_uptime_get_32();
		ret = i2c_burst_read_dt(&i2c_device, io6408_id_register, &value, sizeof(value));
		end_time = k_uptime_get_32();
		if (ret) {
			printk("Failed to read port expander error %d\n", ret);
		}

		if ((end_time - start_time) > 20) {
			printk("i2c_burst_read took %d ms, check logic analyser\n", end_time - start_time);
		}
```

The system is otherwise idle other than this thread, so there shouldn't be anything blocking for 30ms.  I tried to collect CTF trace data, but it only shows a few entries before running into a parse error so it wasn't of any use.

**Reply by EricNRS** [reply]
Was able to exclude the I2C framework from the issue, it looks like this is a system-wide issue.  I wrote some code to just toggle a GPIO pin every 1000 microseconds and sure enough, every few seconds there is a 30ms delay.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/127365838/a63fc21f-4034-4e2a-89ed-61bb8548c8b6)

I turned off cooperative multitasking and adjusted all thread priorities to be lower than the pin-toggle thread and the issue still occurs, so this must be done in an ISR somewhere.  I disabled WiFi and it still continued.  I do have USB attached, so that would be the last possible thing to consider.

**Reply by sylvioalves** [reply]
@EricNRS, can you repeat this test disabling usb console interface and use only uart port?

**Reply by EricNRS** [reply]
@sylvioalves Same problem with USB which was surprising since I have been having issues with USB dropping characters randomly in addition to the I2C failures, so I figured there could be an issue there.  I also recently enabled Modbus and am experiencing timeout failures there as well even though it is only running at 9600 baud.

I will setup a fresh test app and start adding things in until it fails.

**Reply by EricNRS** [reply]
@sylvioalves Success!  Thanks for your help with this!

I setup a semi-automated test to bisect the issue and finally narrowed it down to the tracing framework.  The root cause is the RAM backend for the trace framework where `tracing_backend_ram_output()` is setup to stop logging when the trace buffer is full (see https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/tracing/tracing_backend_ram.c#L26-L29).  I had some local modifications to clear the buffer and restart logging which took 30ms with a 1MB trace buffer.  With `CONFIG_TRACING_SYNC=y`, it does the clearing after calling `irq_lock()`, so that explains the delay.  This also explains why there was never any trace data covering the delay event when I looked at it.  :-/

Disabling  `CONFIG_TRACING_SYNC=y` has resolved the 30ms delays and I have had zero I2C and Modbus failures after several hours of testing.

I'll follow-up with the tracing framework team to see if there are some better options to having the RAM backend use a true circular buffer instead of just stopping when the buffer is full.

**Reply by EricNRS** [reply, marked as answer]
Disabling CONFIG_TRACING_SYNC=y has resolved the 30ms delays and I have had zero I2C (in addition to Modbus and USB) failures after an overnight test run.

**Reply by sylvioalves** [reply]
@EricNRS, Nice to hear that. I didn't see any mention before about `CONFIG_TRACING_SYNC` being enabled. I considered you were using some very basic sample without any additional features. That make sense now. Thanks for bringing all those information. This might be useful in future cases/references.

**Reply by EricNRS** [reply]
@sylvioalves Yes, sorry about that.  I went the other way and disabled everything to create the equivalent to the blinky sample, but I left the TRACING framework on since I wanted to compare traces with the full system.  Luckily the bisect test yesterday took the thinking out of it.
